store.go

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package rate
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+)
+
+type store interface {
+	// fetch will get a Quota for the provided key.
+	// If no quota is found, a new one will be created using the provided Limit.
+	fetch(key string, limit *Limit) (*Quota, error)
+}
+
+type entry struct {
+	key   string
+	value *Quota
+
+	bucket int
+}
+
+type bucket struct {
+	entries map[string]*entry
+
+	expiresAt time.Time
+}
+
+type expirableStore struct {
+	maxSize int
+
+	items map[string]*entry
+
+	buckets            []bucket
+	bucketTTL          time.Duration
+	numBuckets         int
+	nextBucketToExpire int
+
+	mu sync.Mutex
+
+	pool sync.Pool
+}
+
+func newStore(ctx context.Context, maxSize int, maxEntryTTL time.Duration, o ...Option) (store, error) {
+	const op = "rate.newStore"
+
+	opts := getOpts(o...)
+
+	switch {
+	case maxSize <= 0:
+		return nil, fmt.Errorf("%s: max size must be greater than zero: %w", op, ErrInvalidMaxSize)
+	case maxEntryTTL <= 0:
+		return nil, fmt.Errorf("%s: max entry ttl must be greater than zero: %w", op, ErrInvalidParameter)
+	case opts.withNumberBuckets <= 0:
+		return nil, fmt.Errorf("%s: number of buckets must be greater than zero: %w", op, ErrInvalidNumberBuckets)
+	}
```

Comment:
@mgaffney I think I still want do to this validation here, and not in `NewLimiter` for 2 reasons.

1. This validation helps prevent some cases where a store would otherwise panic. It seems safer to do these checks here to ensure the store can be used.
2. As noted [here](https://github.com/hashicorp/go-rate/pull/1/files#diff-c3f1fd6fa2457438e1cd957eea2ba705f9825a4b6e3ad3fa55e3e03be05dea05R65) there is at least one case where this store will not be needed, in which case this validation is not necessary.